vowpalwabbit/parse_example.cc

```diff
@@ -4,17 +4,45 @@
 
 #include <cmath>
 #include <cctype>
+#include <future>
+#include <thread>
+#include <chrono>
 #include "parse_example.h"
 #include "hash.h"
 #include "unique_sort.h"
 #include "global_data.h"
 #include "constant.h"
 #include "vw_string_view.h"
+#include "io_to_queue.h"
+
+class IO_Item;
 
 size_t read_features(vw* all, char*& line, size_t& num_chars)
 {
-  line = nullptr;
-  size_t num_chars_initial = readto(*(all->p->input), line, '\n');
+ 
+  IO_Item result;
+
+  while(true){
+
+    if((*all).p->_io_state.done_with_io && (*all).p->_io_state.io_lines->size() == 0){
+      break;
+    }
+
+    result = pop_io_queue(all);
+
+    if(result.numCharsInit > 0){
+      break;
+    }
+
+  }
+
+  std::string result_string =  result.message;
```

Comment:
To remove a copy you can change this to:
```suggestion
  const auto& result_string =  result.message;
```